### PR TITLE
Implement an `outbound_streams` function

### LINKED ISSUE #654: RFE: Provide a count of open streams
It would be nice to have a count of outbound streams that have not been finished yet.  When this count is zero, the connection can be safely closed.

### PR DESCRIPTION
**Pull request by DemiMarie-parity**
Fixes #654

### CODE REVIEW

**Comment by DemiMarie-parity** [comment]
I did not implement tests because I am not particularly sure what to write.  Any advice would be greatly appreciated.

**Comment by djc** [comment]
I feel like this state should maybe live inside `Streams`? Otherwise looks okay.

File: quinn-proto/src/connection.rs [hunk 1]
```diff
@@ -2598,7 +2603,10 @@ where
     ///
     /// This does not ensure delivery of outstanding data. It is the application's responsibility to
     /// call this only when all important communications have been completed, e.g. by calling
-    /// `finish` on outstanding streams and waiting for the corresponding `StreamFinished` event.
+    /// [`Connection::finish`] on outstanding streams and waiting for the corresponding
+    /// `StreamFinished` event.
+    ///
+    /// It is safe to call this if [`Connection::outbound_streams`] returns 0.
```
Comment (Ralith) [inline quinn-proto/src/connection.rs]: This should be reworded as "safe" has a different very specific meaning in rust.

File: quinn-proto/src/connection.rs [hunk 2]
```diff
@@ -154,6 +154,8 @@ where
     rem_cids: Vec<IssuedCid>,
     /// State of the unreliable datagram extension
     datagrams: DatagramState,
+    /// The number of open *outbound* streams.
+    outbound_streams: u64,
```
Comment (Ralith) [inline quinn-proto/src/connection.rs]: We should use the standard term "send streams" throughout. I also agree that this probably makes more sense as a member of `Streams`.

Comment (DemiMarie-parity) [reply quinn-proto/src/connection.rs]: Agreed.  The difficult part is knowing when to decrement the number of send streams.  I suspect that there are missing calls to `maybe_cleanup`.

File: quinn-proto/src/connection.rs (hunk 1 above)
Comment (DemiMarie-parity) [reply quinn-proto/src/connection.rs]: Done.

**Comment by djc** [comment]
(All of these commits should be squashed before merging.)

**Comment by DemiMarie-parity** [comment]
This may also fix other bugs, in particular a potential memory leak.

**Comment by djc** [comment]
> This may also fix other bugs, in particular a potential memory leak.

Those changes should remain in a separate commit, please.

**Comment by DemiMarie-parity** [comment]
> > This may also fix other bugs, in particular a potential memory leak.
> 
> Those changes should remain in a separate commit, please.

These changes will need to come first, since they are essential for accurate counting of outgoing streams.

**Comment by djc** [comment]
Sure, you can reorder the commits accordingly.

File: quinn-proto/src/connection.rs [hunk 3]
```diff
@@ -910,13 +910,19 @@ where
...
             DataSent { .. } => {
-                self.events.push_back(Event::StreamFinished {
-                    stream: stream_id,
-                    stop_reason,
-                });
                 // No need to hold on to the stop_reason since it's propagated above
-                stream.state = ResetSent { stop_reason: None };
-                self.streams.send_streams -= 1;
+                stream.state = if stopped {
+                    ResetRecvd { stop_reason: None }
```
Comment (Ralith) [inline quinn-proto/src/connection.rs]: This, and the similar change below, is incorrect. `STOP_SENDING` does not constitute acknowledgement of a reset. See [the Stream States](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#name-stream-states) section of the transport draft for details.

Comment (DemiMarie-parity) [reply quinn-proto/src/connection.rs]: Fixed.

**Comment by DemiMarie-parity** [comment]
Does the implementation in this PR guarantee that if `Connection::send_streams()` returns 0, all outstanding data has been acknowledged?

File: quinn-proto/src/connection.rs [hunk 4]
```diff
@@ -909,11 +909,13 @@ where
                 return;
             }
             DataSent { .. } => {
-                self.events.push_back(Event::StreamFinished {
-                    stream: stream_id,
-                    stop_reason,
-                });
                 // No need to hold on to the stop_reason since it's propagated above
+                if stopped {
```
Comment (djc) [inline quinn-proto/src/connection.rs]: This seems to clearly belong in a separate commit.

File: quinn-proto/src/connection.rs [hunk 5]
```diff
@@ -2600,7 +2602,11 @@ where
     ///
     /// This does not ensure delivery of outstanding data. It is the application's responsibility to
     /// call this only when all important communications have been completed, e.g. by calling
-    /// `finish` on outstanding streams and waiting for the corresponding `StreamFinished` event.
+    /// [`Connection::finish`] on outstanding streams and waiting for the corresponding
```
Comment (djc) [inline quinn-proto/src/connection.rs]: Let's also keep documentation-only improvements in another separate commit.

File: quinn-proto/src/connection.rs (hunk 4 above)
Comment (DemiMarie-parity) [reply quinn-proto/src/connection.rs]: Is it correct?

Comment (djc) [reply quinn-proto/src/connection.rs]: I think @Ralith knows better than I do.

File: quinn-proto/src/connection.rs (hunk 5 above)
Comment (DemiMarie-parity) [reply quinn-proto/src/connection.rs]: One reason I changed the documentation is that I could not tell if, should `Connection::write` and/or `Connection::finish` return `WriteError::Stopped`, a `StreamFinished` event will be received.  I assume the answer is “no”, but this should be made clearer.

File: quinn-proto/src/connection.rs (hunk 4 above)
Comment (Ralith) [reply quinn-proto/src/connection.rs]: This doesn't impact the protocol state machine, but changes the quinn-proto API behavior such that if you explicitly reset a stream that you're already finishing, no `StreamFinished` event is yielded. This does not affect quinn, as the interface is constructed to statically forbid that.

I can see the case for the change; `StreamFinished` normally indicates that either the stream was successfully delivered in full or the peer stopped it. I'm a bit concerned that in the hypothetical world where you're finishing and resetting a stream concurrently this doesn't give feedback about the status of the finish operation, but as we don't currently permit that in quinn and a hypothetical quinn-proto user won't necessarily need quinn-proto to communicate that for them, I think the change is reasonable as written--though I agree it belongs in an isolated commit.

File: quinn-proto/src/connection.rs (hunk 5 above)
Comment (Ralith) [reply quinn-proto/src/connection.rs]: I agree, and this is a good change. It's just easier to review logically independent changes for correctness if they're in separate commits, and easier to bisect changes in the future.

File: quinn-proto/src/connection.rs [hunk 6]
```diff
@@ -915,6 +915,7 @@ where
                 });
                 // No need to hold on to the stop_reason since it's propagated above
                 stream.state = ResetSent { stop_reason: None };
+                self.streams.maybe_cleanup(stream_id)
```
Comment (Ralith) [inline quinn-proto/src/connection.rs]: This call, and the similar one added below, isn't needed. Cleanup of reset streams occurs when the reset is acknowledged (see `on_packet_acked`).

File: quinn-proto/src/connection.rs (hunk 4 above)
Comment (Ralith) [reply quinn-proto/src/connection.rs]: nit: Strictly speaking, the commit message should read "locally reset."

Comment (DemiMarie-parity) [reply quinn-proto/src/connection.rs]: I am going to revert this.  Guaranteeing that every successful call to `Connnection::finish` results in a `StreamFinished` event (unless the connection is lost) makes API design easier, and avoids difficult-to-debug hangs.

File: quinn-proto/src/connection.rs (hunk 6 above)
Comment (DemiMarie-parity) [reply quinn-proto/src/connection.rs]: Fixed.

File: quinn-proto/src/streams.rs [hunk 7]
```diff
@@ -283,13 +283,18 @@ impl Send {
...
-    /// If the peer issues additional flow control credit, an `Event::StreamWritable` will be
+    /// If the peer issues additional flow control credit, a [`StreamWritable`] event will be
     /// generated, indicating that retrying the write might succeed.
+    ///
+    /// [`StreamWritable`]: crate::Event::StreamWritable
     #[error(display = "unable to accept further writes")]
     Blocked,
-    /// The peer is no longer accepting data on this stream.
+    /// The peer is no longer accepting data on this stream. No [`StreamFinished`] event will be
+    /// emitted for this stream.
```
Comment (Ralith) [inline quinn-proto/src/streams.rs]: Maybe clearer to say "it's not useful to call `finish`", since `StreamFinished` is not normally generated in association with operations that might return a `WriteError`.

File: quinn-proto/src/connection.rs [hunk 8]
```diff
@@ -2598,7 +2598,13 @@ where
     ///
     /// This does not ensure delivery of outstanding data. It is the application's responsibility to
     /// call this only when all important communications have been completed, e.g. by calling
-    /// `finish` on outstanding streams and waiting for the corresponding `StreamFinished` event.
+    /// [`Connection::finish`] on outstanding streams and waiting for the corresponding
+    /// [`StreamFinished`] event.
+    ///
+    /// If [`Connection::send_streams`] returns 0, all outstanding data has been delivered. There
```
Comment (Ralith) [inline quinn-proto/src/connection.rs]:
```suggestion
    /// If [`Connection::send_streams`] returns 0, all outstanding stream data has been delivered. There
```

e.g. datagrams may not have been delivered.

File: quinn-proto/src/connection.rs (hunk 4 above)
Comment (Ralith) [reply quinn-proto/src/connection.rs]: It might be worth adding (in a future PR) a flag in `StreamFinished` which indicates that the finish "completed" due to a local reset, then, since as it stands it's indistinguishable from success.
